### PR TITLE
Add support for uploading files

### PR DESCRIPTION
**Pull request by yifeikong**

### CODE REVIEW

**Comment by T-256** [comment]
perhaps we could also have `file=` that auto complements multipart?
We should take a look at how other http-clients have implemented `file` parameter.

**Comment by yifeikong** [comment]
I have looked httpx, it's the same as requests. They support files as dict and list of tuple(with length as 2/3/4), which is almost like they hate typehints. I don't think I would do the same thing here. And as I observed, people use helpers like `requests_belt` to upload files.

aiohttp is different, more like ours, but we don't have the `FormData` class, so I won't try to be compatible with them, too.

File: curl_cffi/curl.py [hunk 1]
```diff
@@ -368,3 +371,72 @@ def ws_send(self, payload: bytes, flags: CurlWsFlag = CurlWsFlag.BINARY) -> int:
 
     def ws_close(self):
         self.ws_send(b"", CurlWsFlag.CLOSE)
+
+
+class CurlMime:
+    def __init__(self, curl: Optional[Curl] = None):
+        self._curl = curl if curl else Curl()
```
Comment (T-256) [inline curl_cffi/curl.py:378]:
```suggestion
        self._curl = Curl() if curl is None else curl
```

File: curl_cffi/curl.py [hunk 2]
```diff
@@ -368,3 +371,72 @@ def ws_send(self, payload: bytes, flags: CurlWsFlag = CurlWsFlag.BINARY) -> int:
...
+        self._form = lib.curl_mime_init(self._curl._curl)
+
+    def addpart(
+        self,
+        name: str,
+        *,
+        content_type: Optional[str] = None,
+        filename: Optional[str] = None,
+        local_path: Optional[Union[str, bytes, Path]] = None,
+        data: Optional[bytes] = None,
```
Comment (T-256) [inline curl_cffi/curl.py:388]: IMO, `content` is better choice.

File: curl_cffi/curl.py [hunk 3]
```diff
@@ -368,3 +371,72 @@ def ws_send(self, payload: bytes, flags: CurlWsFlag = CurlWsFlag.BINARY) -> int:
...
+            if ret != 0:
+                raise CurlError("Add field failed.")
+
+        # remote file name
+        if filename is not None:
+            ret = lib.curl_mime_filename(part, filename.encode())
+            if ret != 0:
+                raise CurlError("Add field failed.")
+
+        if local_path and data:
```
Comment (T-256) [inline curl_cffi/curl.py:408]:
```suggestion
        if local_path is not None and data is not None:
```

File: curl_cffi/curl.py [hunk 4]
```diff
@@ -368,3 +371,72 @@ def ws_send(self, payload: bytes, flags: CurlWsFlag = CurlWsFlag.BINARY) -> int:
...
+            raise CurlError("Can not use local_path and data at the same time.")
+
+        # this is a filename
+        if local_path is not None:
```
Comment (T-256) [inline curl_cffi/curl.py:412]:
```suggestion
        elif local_path is not None:
```

File: curl_cffi/curl.py [hunk 5]
```diff
@@ -368,3 +371,72 @@ def ws_send(self, payload: bytes, flags: CurlWsFlag = CurlWsFlag.BINARY) -> int:
...
+            if not isinstance(local_path, bytes):
+                local_path = str(local_path).encode()
+            if not Path(local_path.decode()).exists():
+                raise FileNotFoundError(f"File not found at {local_path}")
+            ret = lib.curl_mime_filedata(part, local_path)
+            if ret != 0:
+                raise CurlError("Add field failed.")
+
+        if data is not None:
```
Comment (T-256) [inline curl_cffi/curl.py:421]:
```suggestion
        elif data is not None:
```

File: curl_cffi/curl.py [hunk 6]
```diff
@@ -368,3 +371,72 @@ def ws_send(self, payload: bytes, flags: CurlWsFlag = CurlWsFlag.BINARY) -> int:
...
+            if not isinstance(data, bytes):
+                data = str(data).encode()
+            ret = lib.curl_mime_data(part, data, len(data))
+
+    @classmethod
+    def from_list(cls, files: List[dict]):
```
Comment (T-256) [inline curl_cffi/curl.py:427]: could we also have `from_dict`?

File: curl_cffi/curl.py [hunk 7]
```diff
@@ -368,3 +371,72 @@ def ws_send(self, payload: bytes, flags: CurlWsFlag = CurlWsFlag.BINARY) -> int:
...
+        form = cls()
+        for file in files:
+            form.addpart(**file)
+        return form
+
+    def attach(self, curl: Optional[Curl] = None):
+        c = curl if curl else self._curl
```
Comment (T-256) [inline curl_cffi/curl.py:434]:
```suggestion
        c = self._curl if curl is None else curl
```

File: curl_cffi/requests/session.py [hunk 8]
```diff
@@ -318,7 +319,14 @@ def _set_curl_options(
 
         # files
         if files:
-            raise NotImplementedError("Files has not been implemented.")
+            raise NotImplementedError("files is not supported, use `multipart`.")
+
+        # multipart
+        if multipart:
```
Comment (T-256) [inline curl_cffi/requests/session.py:325]:
```suggestion
        if multipart is not None:
```

File: curl_cffi/curl.py [hunk 9]
```diff
@@ -368,3 +371,72 @@ def ws_send(self, payload: bytes, flags: CurlWsFlag = CurlWsFlag.BINARY) -> int:
...
+
+        if local_path and data:
+            raise CurlError("Can not use local_path and data at the same time.")
+
+        # this is a filename
+        if local_path is not None:
+            if not isinstance(local_path, bytes):
+                local_path = str(local_path).encode()
+            if not Path(local_path.decode()).exists():
+                raise FileNotFoundError(f"File not found at {local_path}")
```
Comment (T-256) [inline curl_cffi/curl.py:416]: doesn't it need to check for read permission?

**Review by T-256 — APPROVED** [review]
Overall LGTM, just few nit suggestions.

**Pull request merged**